### PR TITLE
Modding new enemies

### PR DESCRIPTION
**Pull request by KABoissonneault**
This is the main part of my efforts to let mods add new enemies to DFU, as discussed here: https://forums.dfworkshop.net/viewtopic.php?f=23&t=4915

This is actually a draft PR, more for raising discussion to see if the solution is appropriate for this point in development.

What this PR does:
- SetupEnemyDemo now recognizes enemy IDs beyond those defined in MobileTypes. For every group of 256 IDs, the first 128 will be EnemyMonster, and the last 128 will be EnemyClass (0-127 Monster, 128-255 Class, 256-383 Monster, 384-511 Class, etc...)
- DaggerfallEntity now has "custom" DFCareer templates, created from mods. This can be used to create new or replace enemy careers for a given enemy ID 
- EnemyEntity will now use that "custom" career template if available, when a DFCareer is needed
- When issued a non-standard enemy ID, TextManager will return the custom career name if available

What this allows:
- With IDs outside the classic values (ie: 0-42 and 128-146), Mod scripts can add new MobileEnemy to `EnemyBasics.Enemies`, and a new DFCareer to `DaggerfallEntity.CustomCareerTemplates`
- Those new enemies can be spawned in three ways: by a quest (through adding the enemy to Quests-Foes), by random encounter (ie: `RandomEncounters.EncounterTables`), and possibly through a Locations editor (not tested)
- With #2160, new enemies would be able to use entirely new graphics

The Mobile will function like a normal enemy, with all the moves normally available to them (nothing special like the seducer transformation, of course). Their name appears when using the Info cursor, and they can be assigned normal equipment through the "OnLootSpawned" event.

A reference implementation for what a mod adding a new enemy type would look like can be found here: https://github.com/KABoissonneault/DFU-NewSprites/blob/main/Scripts/NewSprites.cs

While there is certainly more that could be done to improve this system, I think it works well enough for a first iteration, and I don't think the current limitations prevent future extensions (ex: letting mods change the encounter formula entirely).

I consider this a draft because I'm not too satisfied with some of the technical decisions. After all, I tried to make the code changes minimal, rather than ideal. For example, to decide whether or not the mod wants an EnemyMonster or an EnemyClass, I went with the ID mapping solution, in a way that doesn't break the classic IDs while allowing for virtually infinite expansion. Also, for `CustomCareerTemplate`, it might not be the best solution, though it may be just sufficient.

In conclusion: IK, don't delay your release for this if you want to think about it, but please consider it before the code freeze. I do believe being able to mod new enemies is a basic mod support feature one would expect from a Daggerfall Unity 1.0 release.

### CODE REVIEW

**Comment by ajrb** [comment]
Have added my thoughts @Interkarma. Don't see any issues in my brief testing, looks fine. Over to you.

**Comment by Interkarma** [comment]
Thanks for the review @ajrb! I really appreciate the second set of eyes on this one.

@KABoissonneault I haven't seen any issues created by this PR in core game. Thank you for the light touch overall and working around existing code as much as possible. I'm happy to merge this so that it can be iterated on down the road. I only ask that you hold any further work on this one until after 1.0 - other than fixing any problems that arise from it or very minor tweaks. I'll leave this one open for another day or so in case you want to action anything Hazelnut raised in his review.

This is a really good change and I'm looking forward to seeing how it gets used by the mod community.

File: Assets/Scripts/Game/TextManager.cs [hunk 1]
```diff
@@ -331,15 +332,29 @@ public string[] GetLocalizedTextList(string key, TextCollections collection = Te
...
+                    return enemyNames[43 + enemyID - 128];
+            }
+            // Handle custom enemies
             else
-                return enemyNames[43 + enemyID - 128];
+            {
+                // TODO: maybe go through TextProvider so mods can try to offer localization?
+                if (!DaggerfallEntity.CustomCareerTemplates.TryGetValue(enemyID, out DaggerfallConnect.DFCareer career))
+                {
+                    return "(invalid enemy)";
```
[the remark this answers is not in the document]
Comment (KABoissonneault) [reply Assets/Scripts/Game/TextManager.cs:355]: I added a debug print with the erroneous enemy ID, should help with debugging at least. I even found an issue in R&R:I! (see Discord DMs)

**Comment by ajrb** [comment]
Looks great, thanks @KABoissonneault :)

**Review by ajrb — APPROVED** [review]

**Comment by rossipaolo** [comment]
Thank you for working on this @KABoissonneault, i think it can be merged.

Maybe a possible improvement could be to generate IDs behind the scenes to avoid any chance of collision, but that would be for after 1.0.

**Review by rossipaolo — APPROVED** [review]
